masschallenge/opsworks-web-python
# AC-6025: Make it hard for engineers to accidentally overwrite production database

**Changes introduced in [AC-6025](https://masschallenge.atlassian.net/browse/AC-6025)**:
- add custom recipes to update RDS security groups when an instance is added or removed from an OpsWorks Stack

**How to Test**:
**Test 1**: User cannot log into RDS instances cross-stack
- New RDS security group configurations have been added to the **staging**, **test1**, **test2** and **test4** stacks
- The above means you are now not able to log into a testN RDS instance from a testX/ staging instance, where N != X.
- Previously, this was possible

**Test 2**: Dynamic security group updates
**Setup**:
- Visit a test stack, make sure that the stack settings have the `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY` with values similar to the other test machines
- Ensure the following settings are correctly set
Under **stack settings** the following are set correctly
1. Branch/Revision == AC-6025 (where the new recipes are)
2. Berkshelf version == 3.1.3 (version that allows us to add custom recipes)
3. **Save**

Under **Layers** -> **MassChallenge** -> **Recipes**, the following are present under **Custom Recipes**
1. **opsworks_deploy_python::security-group-update** is under the **Configure** group
2. **opsworks_deploy_python::revoke-security-group** is under the **Shutdown** group
3. **Save**

Under **Deployments** -> **Run Command**
- Select the **Update Custom Cookbooks** command to ensure the instances are using the updated code
- Hit the **Update Custom Cookbooks** button at the bottom to run the command

**Tests**:
1. Existing EC2 instance IP addresses can be added to an RDS security group by executing a recipe on the instance
Under **Deployments** -> **Run Command**
- Select **Execute Recipes** from the **Command** dropdown
- Enter **opsworks_deploy_python::security-group-update** in the **Recipes to execute** textbox
- Hit the **Execute Recipes** button at the bottom of the page
- Once this is done, the **Private IP address** of the existing instances will be in an RDS security group associated with the master RDS instance of that stack

2. A new instance is automatically added to the master RDS security group
- Under **instances**, click the link to **+ instance**
- Select the **m3.large** size
- Click the **Add instance** button
- It will take a while for it to completely setup, but at the end you will notice that the **Private IP address** of the instance has been added to an **RDS security group** associated with the master RDS instance that the stack uses

3. Stopping an instance has it's privileges revoked
- Under **instances**, hit the **Stop** link under the **Actions** column
- Once it finishes shutting down, you will notice the **Private IP address** of the stopped instance is no longer a part of the authorised IPs under the **RDS security groups** of the associated master RDS instance of the stack.

**NOTE**:
- If this PR is merged to master [as is](https://github.com/masschallenge/opsworks-web-python/commit/92ca19c82481328412e46fcb7f7497879a6a5dba), we have to give new permissions to the 
[accelerate-s3-production-manager](https://console.aws.amazon.com/iam/home?region=us-east-1#/users/accelerate-s3-production-manager) user for the recipes to work in Production.
